File: 2025/day3/day3.nim
import strutils, sequtils, math

proc combos(digits: seq[int], k: int): seq[seq[int]] =
  if k == 0: return @[@[]]
  if digits.len == 0: return @[]
  let rest = digits[1..^1]
  for c in combos(rest, k - 1): result.add(@[digits[0]] & c)
  result.add combos(rest, k)

proc bestNumber(line: string, count: int): int =
  let digits = line.toSeq.filterIt(it.isDigit).mapIt(it.ord - '0'.ord)
  for combo in combos(digits, count):
    let val = combo.foldl(a * 10 + b, 0)
    result = max(result, val)

let lines = readFile("inputs.txt").strip.splitLines
echo "Part 1: ", lines.mapIt(bestNumber(it, 2)).sum
echo "Part 2: ", lines.mapIt(bestNumber(it, 12)).sum

File: 2025/day3/day3.py
def best_number(line, k):
    digits = [int(ch) for ch in line if ch.isdigit()]
    drop = len(digits) - k
    stack = []

    for d in digits:
        while drop > 0 and stack and stack[-1] < d:
            stack.pop()
            drop -= 1
        stack.append(d)

    # take first k digits
    stack = stack[:k]

    val = 0
    for d in stack:
        val = val * 10 + d
    return val


p1 = p2 = 0

with open("inputs.txt") as f:
    for line in f:
        line = line.strip()
        p1 += best_number(line, 2)
        p2 += best_number(line, 12)

print("Part 1:", p1)
print("Part 2:", p2)